### ISSUE TITLE
[SuperEditor] markdown support for bold and italics 

### ISSUE THREAD

**Issue by miguelcmedeiros** [superlistapp/super_editor#1866]
Support edit reactions to detect markdown annotation for bold and italics:
* when the user types `something *italic*` it should convert the content to "something *italic*"
* when the user types `something **bold**` it should convert the content to "something **bold**"

**Comment by miguelcmedeiros** [comment]
Cc @matthew-carroll

**Comment by matthew-carroll** [comment]
@miguelcmedeiros I'd like to flesh out the intended/expected behavior for this ticket. Here are some questions:

When the user is in the middle of typing markdown styles, should the styles be applied pre-emptively before the style is complete? E.g., `"This is partially **bold"` and we bold the "bold" text without a closing marker.

Do the styles stick around after editing? E.g., does `"This is **bold**" ` become "This is **bold**", or do the "*"s remain? 

Based on the above question, if the style syntax disappears after editing, does it come back when the user places the caret inside the styled text? Or is it a fire and forget situation where you apply the Markdown, it generates the styles, and then it's gone forever?

When the user selects some text and presses CMD+B to bold it, does that insert Markdown around the text? Or does it continue to apply the bold style as usual?

**Comment by miguelcmedeiros** [comment]
From other tools like Notion and Linear, it's fire and forget, i.e. the style syntax is removed and the appropriate attribution is applied.

**Comment by matthew-carroll** [comment]
@miguelcmedeiros to limit complexity, I'm working on a solution that only looks for Markdown tokens within a single node at a time. So, for example, "**" can't be used to bold text at the end of one paragraph into the beginning of the following paragraph.

Is that limitation acceptable for your use, or do you need styling to span nodes?

Also, to avoid re-inventing Markdown parsing for inline styles, I'm taking any modified `TextNode` and passing it through the standard Markdown parser. Then, I'm comparing the parsed text. If the text is different after being parsed, it means some piece of it was Markdown and was applied to the text. Then I'm replacing the previous `TextNode` content with the new content.

As a result, when the Markdown style is applied, it won't look like a surgical edit. Instead, it will look like the whole paragraph is replaced with another paragraph. Do you know if this level of replacement would be a problem for Superlist?

**Comment by dhikshith12** [comment]
Hey @matthew-carroll, It is interesting to know your approach, I've tried to do it with a new implementation of `EditReaction` and adding it to `defaultEditReactions`.

But this would mean, as you mentioned "re-inventing Markdown parsing" for inline styles, your way will be a more general solution that will work for any other type of inline styles apart from bold and italic.

I was trying to just solve these three cases 
- italic - `*italic*`
- bold - `**bold**`
- bold and italic - `***bold and italic***`

The first time I tried I got the deletion working, i.e., replace `**bold**` with `bold`, and looked into styling to complete the other part of actually replacing it with correct Attributions, some concepts were new to me, overall I'd say this is a well-designed & well-architected project from an outside look (I'm new here).

I'll let you know if I made any progress on this 👍

**Comment by matthew-carroll** [comment]
@miguelcmedeiros @dhikshith12 - Here are a couple hurdles I just ran into with my described approach.

First, at least by default, the Markdown package will parse something like "**bold*" -> "*bold" with italics. I.e., the parser doesn't look/wait for balanced tokens. Therefore, as the user types "**bold*", before the user can get to "**bold**", it's already parsed as italics. Perhaps there's an option in the Markdown package to prevent this, but I don't know what it is.

Second, adjusting the user's selection in response to the Markdown change might be tricky. For example, when the user types "**bold**" and it's converted to "bold", the text content went from 8 characters down to 4 characters, so the user's caret needs to move from position 8 to position 4. Furthermore, not every token is 2 characters long. Italics "*" and strikethrough "~" are one character long. Link notation is variable length. And it's possible that the user pasted some Markdown content, which means more than a single token might be parsed and replaced at a given moment in time. Therefore, by parsing whole-sale, it's not clear to me how to know where the user's caret or expanded selection should end up.

If we parsed token-by-token ourselves, then in theory we could correct for both of these issues. However, parsing the tokens ourselves means implementing the pattern recognition for all the inline syntaxes: "**", "__", "*", "_", "~", "[]()", "``".

It also means handling mixed matches and adding intelligence to handle unbalanced cases, e.g. "_*bold*_", "**bold*".

At the moment I'm not sure which one is more work to implement correctly. I think I might press forward a bit further with the whole-sale replacement using the built in Markdown parser. At least in that case, if I'm able to hack together something that works, we can fill out the test suite and then replace it with a better solution.

**Comment by dhikshith12** [comment]
To answer these questions, it would be helpful to think about where we want `super_editor` to be on the markdown support spectrum, on the one side we have Obsidian which is a pure Markdown editor, and Notion on the other side, which is not a pure-Markdown editor, but has Syntax styling support for those that have a habit of using markdown syntax for styling while typing.

Your approach is leaning more toward Obsidian pure Markdown style, while a Manual Syntax Conversion is leaning toward Notion.

While exploring the repo, i found how `## Heading` input is handled to convert to proper Heading level, and it's done using an EditReaction as of now, and that was the inspiration to my approach, 

If your approach works at a generalisable level we can remove how heading input and a few other manually handled syntax to delegate it to Markdown package, because otherwise they may collide trying to manipulate the document at once.

**Comment by matthew-carroll** [comment]
> it would be helpful to think about where we want super_editor to be on the markdown support spectrum

Right now we're focused on whatever Superlist requests because they're funding this feature.

Also, Super Editor should make it possible to achieve whatever position on that spectrum you want. The only question is what ships with the Super Editor package.

> If your approach works at a generalisable level we can remove how heading input

As I mentioned above, my approach doesn't do anything at the overall Markdown block level. It probably could, if extended, but based on what I've discussed so far, my approach wouldn't have any impact on block-level parsing.

**Comment by matthew-carroll** [comment]
I was able to get nominal selection updates, which made it possible to demo the feature. However, I quickly realized that if we're deleting the whole paragraph and replacing it, we need to explicitly retain the original paragraph's attributions so we don't lose formatting along the way. Those attributions can't just be copied, because their indices come from the original text with Markdown characters. I need to repeat the same update I just did for the selection location to each of the attribution markers.


Uploading Screen Recording 2024-03-08 at 11.35.04 PM.mov…

**Comment by matthew-carroll** [comment]
I brought in a text diff'ing package and got the reaction to the point where it retains existing attributions and adds new attributions.


https://github.com/superlistapp/super_editor/assets/7259036/be9e6175-c2c3-4898-8e63-7008b564a19c

@miguelcmedeiros we could us some direction here...

Is it acceptable for the reaction to only apply to individual paragraphs, and ignore any possible Markdown that spans across consecutive paragraphs?

Is it acceptable for this reaction to completely replace the previous paragraph with a new paragraph after finding any new Markdown syntax? Or do we need to calculate the insertions and deletions so that the change log is precise?

I still need to figure out how to prevent `**bold*` from becoming **italics*.

**Comment by matthew-carroll** [comment]
I found a workaround for applying bold. I now take the original paragraph `AttributedText` and serialize that to Markdown. Then, I deserialize back to `AttributedText`. This gives us the effect we want.


https://github.com/superlistapp/super_editor/assets/7259036/b5de6fd1-b1df-4c66-94f7-82a732257b91

**Comment by miguelcmedeiros** [comment]
@matthew-carroll missed this ticket. Sorry about that.

I tested it out today in the demo app and found the following:
* `**bold*` becoming italic is indeed weird to have. From the current PR I guess it's not solved yet.
* I can't type a `*` as the first character of a paragraph, simply does nothing. Without enabling any of the existing loggers, I get the following from the print statements:
```
flutter: Replacing markdown with attributions.
flutter:  - current selection: [DocumentSelection] - 
  base: ([DocumentPosition] - node: "ef78e3c0-e622-4378-8bea-3db0dc5de271", position: (TextPosition(offset: 1, affinity: TextAffinity.downstream))),
  extent: ([DocumentPosition] - node: "ef78e3c0-e622-4378-8bea-3db0dc5de271", position: (TextPosition(offset: 1, affinity: TextAffinity.downstream)))
flutter: Checking edited node for Markdown changes: ef78e3c0-e622-4378-8bea-3db0dc5de271
flutter: 
flutter: 
flutter: Processing inline Markdown diffs:
flutter: DELETE: *
flutter: Creating delete request from 0 to 1
flutter: Updated extent to 0
flutter: Original: '*'
flutter: Parsed: ''
flutter: 
flutter: 
flutter: Delta change requests:
flutter: Will be applied to: '*'
flutter: Insertions and deletions:
flutter:  - delete [DocumentRange] - start: ([DocumentPosition] - node: "ef78e3c0-e622-4378-8bea-3db0dc5de271", position: (TextPosition(offset: 0, affinity: TextAffinity.downstream))), end: ([DocumentPosition] - node: "ef78e3c0-e622-4378-8bea-3db0dc5de271", position: (TextPosition(offset: 1, affinity: TextAffinity.downstream)))
flutter: 
flutter: --------
```

**Comment by matthew-carroll** [comment]
@miguelcmedeiros 

> `**bold*` becoming italic is indeed weird to have. From the current PR I guess it's not solved yet.

What would be the solution? If the user types `**something* and more stuff here`, what's the expected or desired result?

If the desire is to recognize this special arrangement of syntax and prevent the application of Markdown, it might be pretty complicated to do that. Because we may need to deeply intervene in the middle of the Markdown parsing package. Is this a critical change to make?

> I can't type a * as the first character of a paragraph, simply does nothing.

I'll look into this and fix it.

**Comment by miguelcmedeiros** [comment]
@matthew-carroll 
> What would be the solution? If the user types `**something* and more stuff here`, what's the expected or desired result?

If the types the text above it I would expect that no attribution is applied since the markdown markers around "something" are different. I see that the current implementation gives similar outcome as we get in GitHub comments, but in the case of GitHub the attribution is not being applied as the user types, but done in a single pass when the raw text is parsed before rendering the final result.

If you can check how Notion handles it, it's a good example of what I would expect as the behavior that the user gets while typing:

https://github.com/superlistapp/super_editor/assets/31278849/10aeb89c-a189-4a60-92f6-eabe9085bf5a

Notion seems to apply an attribution in the following cases:
* `**word**`
* `some **word**`
* `long **bold**word`

It won't apply the attribution in the following cases:
* `long**word**withoutspaces`
* `**word*`
* `**word *`

**Comment by matthew-carroll** [comment]
I'm going to try to implement this desired behavior, but for the record, I filed this issue with the `markdown` package: https://github.com/dart-lang/markdown/issues/599

**Comment by matthew-carroll** [comment]
@miguelcmedeiros what should happen with `**this* and *that**`?

**Comment by miguelcmedeiros** [comment]
@matthew-carroll personally think that the Linear/Notion behavior is easier to understand/follow as the user types.

> I think their policy might be to look only at the token touching the caret.

I think it also tries to match the token touching the caret with a matching token preceded by a space character or at the start of a line.

**Comment by matthew-carroll** [comment]
@miguelcmedeiros I pushed a rework to #1896 - I retained the work on full deserialization but also added another reaction that only considers the location of the caret, and only walks upstream from the caret.

This new reaction is essentially a partial Markdown parser. I think I have the reaction doing what Notion does with inline Markdown.


https://github.com/superlistapp/super_editor/assets/7259036/016ce9f6-1bac-4e34-9314-11980a7e7088

I noticed that Notion doesn't serialize Markdown links when the user types. Are you OK with us not serializing those?
